### ISSUE TITLE
Way to block managed resource from being deleted till its dependency deleted

### ISSUE THREAD

**Issue by morningspace** [crossplane/crossplane#2439]
### What problem are you facing?

I have two managed resources where one depends on the other. When delete the two resources, one will fail if the other is deleted too early. Right now, it looks the general reconciler handles managed resources deletion independently, which is good. The thing is that wether it could allow provider owner to customize this behavior somehow using `finalizer`.

Couple of observations after investigation for a while:
- Some provider implementation are kind of taking over the reconciliation logic by themselves, so that they fully manage the finalizer adding/removing by themselves. e.g.: the [storage/container controller](https://github.com/crossplane/provider-azure/blob/master/pkg/controller/storage/container/container.go) in provider-azure. But this will lose all benefits provided by the general managed resource reconciler, e.g. its Connect/Observer/Create/Update/Delete lifecycle.
- Inside the provider, It looks I can handle the finalizer of the managed resource passed in, but it looks the current implementation for the general reconciler assumes there is only one finalizer, `finalizer.managedresource.crossplane.io`, which is added by itself. This may lead to the confusion even multi-finalizer approach is technically feasible. e.g.: after the reconciler remove its finalizer from the managed resource, it will report "Successfully deleted managed resource".

### How could Crossplane help solve your problem?

It looks the scenario itself is something common, but maybe I'm too opinionated:-). It would be great if there's a nice way to handle it in crossplane core or runtime, but if that's something not too generic, maybe it makes sense to keep that inside the particular provider at least for now, in which case, I think provider owners may need some guidance on this from the community owner.

**Comment by morningspace** [comment]
Seeing a similar issue reported here: https://github.com/crossplane-contrib/provider-sql/issues/47, so I see more people having this similar issue. It could be fixed somehow at provider level, but given more providers potentially having this problem, it would be better to address it properly at crossplane core & runtime level.

**Comment by negz** [comment]
> I have two managed resources where one depends on the other.

Can you give an example of the nature of the dependency?

> Some provider implementation are kind of taking over the reconciliation logic by themselves, so that they fully manage the finalizer adding/removing by themselves. e.g.: the storage/container controller in provider-azure.

This is an anti-pattern that will be removed. The provider-azure storage controllers in fact predate the `managed.Reconciler`, so the issue is more that they have not yet been updated rather than that they're doing something special on purpose.

> It could be fixed somehow at provider level, but given more providers potentially having this problem, it would be better to address it properly at crossplane core & runtime level.

We have consciously avoided this level of dependency modelling in Crossplane - i.e. having Crossplane create, delete, or update things in an ordered manner. Doing so is not idiomatic for Kubernetes and would increase the complexity of our logic significantly. Instead we typically prefer for each `ExternalClient` to tolerate eventual consistency. For example I suspect the right fix for https://github.com/crossplane-contrib/provider-sql/issues/47 is for the Grant ExternalClient to handle the case where the User no longer exists during deletion, rather than creating a dependency on the two at the Kubernetes level such that a User cannot be deleted until its Grants are first deleted.

**Comment by morningspace** [comment]
@negz To answer your question
> Can you give an example of the nature of the dependency?

My scenario actually comes from the use of [ODLM](https://github.com/IBM/operand-deployment-lifecycle-manager/), which is sort of workload launch abstraction on top of [OLM](https://github.com/operator-framework/operator-lifecycle-manager) and can be used to provision operators and operands (the actual workloads) either in local or remote cluster, depending on kubeconfig. To use it, I need to create two resources called `OperandRegistry` and `OperandConfig` for some preparation, then a third resource called `OperandRequest` to request the actual operators and operands that I want to launch. For uninstall in my case, `OperandRegistry`, `OperandConfig`, `OperandRequest` are all Managed Resources, which will be deleted independently. But since `OperandRequest` depends on the other two resources, sometimes it will fail to be deleted because its dependencies are deleted too early.

> This is an anti-pattern that will be removed.

Got it, thanks for clarification. I did not realize that the provider-azure storage controllers implementation is out-of-date :-)

> We have consciously avoided this level of dependency modeling in Crossplane ...

I can understand your point. I see there's an ongoing discussion at https://github.com/crossplane/crossplane/issues/2072, where people were talking about the creation order which still remains open. I think this issue is related, just on the opposite side, i.e. create or install vs. delete or uninstall. I see your suggestion that relies on `ExternalClient ` to tolerate eventual consistency. Actually I've opened [an issue](https://github.com/IBM/operand-deployment-lifecycle-manager/issues/714) against ODLM team on this. Before that, in order to unblock, I am planning to address it at provider level. So, it does not impact crossplane core or runtime, which is good, and also, it can complete the e2e install/uninstall flow.

**Comment by natereid72** [comment]
To add additional consensus. If you delete a resource that created an AWS load balancer (e.g. release), and an EKS cluster at the same time, if the order of those two resources aren't deleted correctly, the load balancer becomes stuck and the VPC can't be deleted until you go into the aws console and manually delete the lb.

**Comment by bobh66** [comment]
What @natereid72 said.  We hang LBs in AWS all of the time and it's a pain.

Another scenario is documented here https://github.com/crossplane-contrib/provider-terraform/issues/47 - I have a Composition that creates a VPC via provider-aws and then does some configuration of the VPC via provider-terraform, since the AWS provider doesn't support the fine-grained configuration that is required.  On delete, if the VPC gets destroyed before the Workspace, the terraform destroy command will fail because it can't find the VPC, and the Workspace is effectively stuck in the deleting state.  I am updating provider-terraform to support a "destroyArgs" parameter that will allow me to have terraform destroy called with the "-refresh=false" argument, so that terraform won't try to find the VPC that is gone and the delete of the Workspace will succeed.

I can also envision much larger compositions where we deploy an "application controller" and then use it to deploy applications, and on the delete side we will need to handle the case where the "application controller" has been deleted before the associated applications.

I'm not making an argument for dependency support, as I agree there should be better ways to do it, but there may well be cases where there is no optimal/possible solution outside of Crossplane.

**Comment by bobh66** [comment]
It seems like this issue could be mitigated by allowing the composition to arrange for owner dependencies between resources within the composition and setting blockOwnerDeletion along with the use of the "forward" cascade policy for delete.  That would potentially allow the user to control the order of resource deletion within the Composition without any changes in Crossplane.

I attempted to see if this would work but it is explicitly blocked here:

https://github.com/crossplane/crossplane/blame/master/internal/controller/apiextensions/composite/composed.go#L366

I can understand wanting to make sure that the Crossplane Composite resource is the "controlling" owner, but I wonder if it would be sufficient to sanitize any existing ownerReferences from the Composition and force controller=false?

@negz - are there any concerns about the Composition providing ownerReferences beyond making sure that the Composite resource is also the controller?

The ability to control deletion dependencies is becoming critical to us and this seems like a reasonable simple and elegant solution, assuming it works.

**Comment by morningspace** [comment]
As a general option, I'm thinking if k8s provider could help somehow in this case, given that something running inside k8s cluster is responsible for the CRUD of the actual remote resource, either directly or indirectly, e.g.: certain provider w/ its custom resource as the remote representation of some service on public cloud. We could create a special type of `Object` w/ [`ObserveDelete` policy](https://github.com/crossplane-contrib/provider-kubernetes/blob/main/apis/object/v1alpha1/types.go#L42) along w/ [references](https://github.com/crossplane-contrib/provider-kubernetes/blob/main/apis/object/v1alpha1/types.go#L84-L98) to the dependent resources (may also need to be mapped to corresponding `Object`s if they are remote). This way, the deletion will happen in an order that dependent resources first, then owner resource, all"controlled" by `Object`s.

**Comment by bobh66** [comment]
The dependency of the Object on the Cluster is rather straightforward other than the fact that there are multiple implementations of a Cluster.  In the AWS case, there is the Cluster.eks object and associated connection Secret which need to stay in existence until all of the dependent Objects/Releases/Workspaces are finished using the K8S API.   There are equivalent dependencies in GCP, Azure, etc.

The helm Release case is more complex as there may be pre/post-delete jobs that need to run before the Helm chart can be successfully deleted, so in addition to the K8S API we also need the associated Nodegroups to stay in service so there is compute available to run the required Jobs.  There may be other related resources that need to stay as well, for example LaunchTemplates, AutoScalingGroups, etc.

Another option might be depth-first and reverse-order delete handling - since the claim is effectively the root of a DAG, an implementation of delete handling via finalizer and using blockOwnerDeletion could descend to the deepest level in the graph and start the deletion process there, processing the list of resources in each composition in reverse order, and working up the tree until everything has been deleted.

There will always be layers of resources and dependencies between layers which can be resolved via reconciliation on creation but which are not easily reversible for deletion.  How much work do we want Crossplane to do to handle this vs allowing the user to specify the dependencies?  For example if the user could specify ownerReferences with blockOwnerDeletion in the managed resources and Crossplane delayed delete handling until the forwardDeletion finalizer is removed, the user would have direct control over the delete processing.

**Comment by valorl** [comment]
Regarding the EKS LB issue, I think that's painful regardless whether the tool you use has the ability to enforce order/dependencies. It should really be handled by EKS. https://github.com/aws/containers-roadmap/issues/1348

**Comment by negz** [comment]
@valorl I do like the idea that EKS should clean up after itself, but unfortunately this only seems to be the most common/painful instance of the problem. I note for example that per https://cloud.google.com/kubernetes-engine/docs/how-to/deleting-a-cluster GKE won't necessarily clean up LBs or volumes when you delete the cluster.

**Comment by muvaf** [comment]
Just noting here that there is some discussion regarding this problem here https://github.com/crossplane/crossplane/discussions/3116#discussioncomment-2856158 .

IMO, `ConnectionSecretUsage` pattern suggested there would cover a lot of grounds here without any user involvement and it's a familiar pattern. But that wouldn't include the cases like LB deletion because Crossplane effectively has no idea or insight into the remote cluster unless they were provisioned by a Crossplane provider, which would make it be covered by `ConnectionSecretUsage` since the provider has to use a connection secret to execute the operations.

**Comment by jaylevin** [comment]
Having this issue after upjetting the Confluent terraform provider. The `Cluster` and `APIKey` MRs are deleted before the `KafkaACL`, resulting in refresh failure because the `KafkaACL` depends on the secret generated by the APIKey.

```
cannot run refresh: refresh failed: expected "credentials.0.secret" to not be an empty string, got : expected "credentials.0.key" to not be an empty string, got :
```

**Comment by nstogner** [comment]
```
Regarding the EKS LB issue, I think that's painful regardless whether the tool you use has the ability to enforce order/dependencies. It should really be handled by EKS.
```

```
I do like the idea that EKS should clean up after itself
```

@valorl @negz In regards to EKS/GKE/etc cleaning up after themselves... I don't think this will ever occur. A kubernetes cluster does not have a bounded set of things it could possibly create. For example: consider a k8s cluster with crossplane installed provisioning another k8s cluster with crossplane installed... I don't see cloud providers attempting to delete Composites/Manage-Resources and wait for their deletion before proceeding with cluster deletion. This case plays out for any k8s extension that could possibly create external resources (ExternalDNS, ACK, GCP Config Connector, and Crossplane to name a few).

**Comment by nstogner** [comment]
@negz I am currently trying to tackle the EKS/GKE cleanup issue. Current direction I am considering:

Implement a vaidating/mutating webhook that intercepts the Deletion request to the Composite.

Alternatively, if Crossplane relies upon active deletion (does it?), perhaps the Crossplane project could add support for a well-known finalizer naming convention on Composites (i.e. `pre-delete.crossplane.io/my-custom-name`). The presence of these finalizer would prevent the deletion of MR until some external controller does it's use-case specific cleanup logic and removes the finalizer.

For more context: I have a controller written that connects to the child cluster, creates a gatekeeper policy to block the creation of new resources, and then deletes (and waits for deletion of) all resources that triggered external resources to be created... delegating the cleanup to the controllers in the child cluster. I am looking for the proper place to hook into Crossplane to trigger this flow.

**Comment by bobh66** [comment]
I have a Kyverno implementation of delete dependency tracking that uses policies and webhooks to keep track of specified dependencies and block the depended-on resources until the dependencies have been deleted.  I have a writeup of the implementation that I will try to get posted this week.

**Comment by nstogner** [comment]
@bobh66 Would love to take a look! I am guessing that would not tackle this use-case though? https://github.com/crossplane/crossplane/issues/2439#issuecomment-1113504045

**Comment by jcooklin** [comment]
> I have a Kyverno implementation of delete dependency tracking that uses policies and webhooks to keep track of specified dependencies and block the depended-on resources until the dependencies have been deleted. I have a writeup of the implementation that I will try to get posted this week.

Same.  I would be interested in seeing what you came up with.

**Comment by bobh66** [comment]
I posted a writeup here - https://github.com/crossplane/crossplane/discussions/4072 - let me know if you have any questions.

**Comment by turkenh** [comment]
Closing as a duplicate of https://github.com/crossplane/crossplane/issues/3393
